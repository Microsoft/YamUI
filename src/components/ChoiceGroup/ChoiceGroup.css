/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import "src/css/variables/colors.css";

.y-choice-group .ms-ChoiceField {
  margin-bottom: 1.2rem;
  min-height: 0;
}

.y-choice-group .ms-ChoiceField-field--image {
  border-width: 1px;
  border-color: #dde0e6;
  background-color: white;
  padding-top: 28px;
}

.y-choice-group .ms-ChoiceField--image {
  margin-right: 1.2rem;
}


