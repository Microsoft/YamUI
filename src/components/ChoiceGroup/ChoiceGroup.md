### Examples

```js { "props": { "data-example": "no selected key" } }
const options = [
  { key: 'A', text: 'Option A', label: 'optional sublabel for A' },
  { key: 'B', text: 'Option B' },
  { key: 'C', text: 'Option C' },
];

<ChoiceGroup label="No Selected Key Provided" options={options} onChange={console.log} />;
```

```js { "props": { "data-example": "selected key provided" } }
const options = [
  { key: 'A', text: 'Option A', label: 'optional sublabel for A' },
  { key: 'B', text: 'Option B' },
  { key: 'C', text: 'Option C' },
];

<ChoiceGroup label="Selected Key Provided" options={options} onChange={console.log} selectedKey="B" />;
```
