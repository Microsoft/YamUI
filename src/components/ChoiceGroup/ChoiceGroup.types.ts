/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { BaseComponentProps } from '../../util/BaseComponent/props';

export interface ChoiceGroupOption {
  /**
   * A required key to uniquely identify the option.
   */
  key: string;

  /**
   * Text for the option.
   */
  text: string;

  /**
   * Label text for the option.
   */
  label?: string;
}

export interface ChoiceGroupProps extends BaseComponentProps {
  /**
   * Items to be presented to the user.
   */
  options: ChoiceGroupOption[];

  /**
   * Label text that will be rendered above the choices.
   */
  label?: string;

  /**
   * The selected option.
   */
  selectedKey?: ChoiceGroupOption['key'];

  /**
   * Function to call when user changes the selected item.
   */
  onChange(key: ChoiceGroupOption['key']): void;
}
