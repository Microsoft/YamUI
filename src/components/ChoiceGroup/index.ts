/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export { default } from './ChoiceGroup';
export * from './ChoiceGroup.types';
