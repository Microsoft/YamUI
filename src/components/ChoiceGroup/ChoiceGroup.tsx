/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import { ChoiceGroup as FabricChoiceGroup, IChoiceGroupOption } from 'office-ui-fabric-react/lib/ChoiceGroup';
import { join } from '../../util/classNames';
import Text from '../Text';
import './ChoiceGroup.css';

export interface ChoiceGroupOption {
  /**
   * A required key to uniquely identify the option.
   */
  key: string;

  /**
   * Text for the option.
   */
  text: string;

  /**
   * Label text for the option.
   */
  label?: string;
}

export interface ChoiceGroupProps extends BaseComponentProps {
  /**
   * Items to be presented to the user.
   */
  options: ChoiceGroupOption[];

  /**
   * Label text that will be rendered above the choices.
   */
  label?: string;

  /**
   * The selected option.
   */
  selectedKey?: ChoiceGroupOption['key'];

  /**
   * Function to call when user changes the selected item.
   */
  onChange(key: ChoiceGroupOption['key']): void;
}

const getTextWithLabel = (label: string, sublabel: string) => {
  return () => (
    <span className="ms-Label">
      <Text bold={true}>{`${label}:`}</Text>
      <Text>&nbsp;{sublabel}</Text>
    </span>
  );
};

const mapOption = (option: ChoiceGroupOption) => {
  const { text, label } = option;
  const onRenderLabel = label ? getTextWithLabel(text, label) : undefined;

  return {
    key: option.key,
    text: text,
    onRenderLabel,
  };
};

/**
 * The ChoiceGroup component, also known as radio buttons, let users select one
 * option from two or more choices. Each option is represented by one ChoiceGroup
 * button; a user can select only one ChoiceGroup in a button group.
 */
export default class ChoiceGroup extends React.Component<ChoiceGroupProps> {
  public render() {
    const { label, className, options, selectedKey } = this.props;
    return (
      <FabricChoiceGroup
        label={label}
        className={join(['y-choice-group', className])}
        options={options.map(mapOption)}
        onChange={this.onChange}
        selectedKey={selectedKey}
      />
    );
  }

  private onChange = (_: any, option: IChoiceGroupOption | undefined) => {
    if (option) {
      this.props.onChange(option.key);
    }
  };
}
