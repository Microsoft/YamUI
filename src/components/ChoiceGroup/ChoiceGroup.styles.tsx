/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { IChoiceGroupStyles, IChoiceGroupStyleProps } from 'office-ui-fabric-react/lib/components/ChoiceGroup/';
import {
  IChoiceGroupOptionStyles,
  IChoiceGroupOptionStyleProps,
} from 'office-ui-fabric-react/lib/components/ChoiceGroup/ChoiceGroupOption/';
import { fontWeights } from '../../util/styles/fonts';

export const getStyles = (props: IChoiceGroupStyleProps): IChoiceGroupStyles => {
  const { theme } = props;
  return {
    label: {
      color: theme.semanticColors.bodyText,
      fontSize: theme.fonts.smallPlus.fontSize,
      fontWeight: fontWeights.bold,
      paddingBottom: '0.4rem',
    },
  };
};

export const getOptionStyles = (props: IChoiceGroupOptionStyleProps): IChoiceGroupOptionStyles => {
  const { focused, checked, theme } = props;
  return {
    root: {
      marginBottom: '1.2rem',
      minHeight: 0,
    },
    field: {
      paddingLeft: '2.6rem',
      selectors: {
        ':before': {
          borderColor: checked || focused ? theme.palette.themePrimary : theme.palette.neutralDark,
          borderWidth: checked ? '2px' : '1px',
        },
      },
    },
  };
};
