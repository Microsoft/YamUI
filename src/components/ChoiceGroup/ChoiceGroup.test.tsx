/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import ChoiceGroup, { ChoiceGroupProps } from '.';
import { IChoiceGroupOption } from 'office-ui-fabric-react/lib/ChoiceGroup';

describe('<ChoiceGroup />', () => {
  let onChange: jest.Mock<Function>;

  const getChoiceGroup = (overrides?: Partial<ChoiceGroupProps>) => {
    onChange = jest.fn();
    const props = {
      onChange,
      options: [
        {
          key: 'A',
          text: 'Option A',
          label: 'Sublabel for A',
        },
        {
          key: 'B',
          text: 'Option B',
        },
      ],
      ...overrides,
    };
    return <ChoiceGroup {...props} />;
  };

  let component: ShallowWrapper<ChoiceGroupProps>;
  let labelWithSublabel: JSX.Element;

  describe('with minimal props', () => {
    beforeEach(() => {
      component = shallow(getChoiceGroup());
      const options = component.find('ChoiceGroup').prop('options');
      const onRenderLabel = (options as IChoiceGroupOption[])[0].onRenderLabel;
      labelWithSublabel = (onRenderLabel as Function)();
    });

    it('renders as expected', () => {
      expect(component).toMatchSnapshot();
    });

    it('sublabel renders as expected', () => {
      expect(labelWithSublabel).toMatchSnapshot();
    });

    describe('when a change is triggered with an option', () => {
      beforeEach(() => {
        component.simulate('change', null, { key: 'B' });
      });

      it('renders as expected', () => {
        expect(component).toMatchSnapshot();
      });

      it('calls `props.onChange`', () => {
        expect(onChange).toBeCalledWith('B');
      });
    });

    describe('when a change is triggered with an option that is `undefined`', () => {
      beforeEach(() => {
        component.simulate('change', null, undefined);
      });

      it('renders as expected', () => {
        expect(component).toMatchSnapshot();
      });

      it('does not call `props.onChange`', () => {
        expect(onChange).not.toBeCalled();
      });
    });
  });

  describe('when `props.selectedKey` is set', () => {
    beforeEach(() => {
      component = shallow(getChoiceGroup({ selectedKey: 'B' }));
    });

    it('renders as expected', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
