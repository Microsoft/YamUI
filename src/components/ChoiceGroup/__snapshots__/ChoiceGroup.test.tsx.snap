// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChoiceGroup /> when \`props.selectedKey\` is set renders as expected 1`] = `
<ChoiceGroup
  className="y-choice-group"
  onChange={[Function]}
  options={
    Array [
      Object {
        "key": "A",
        "onRenderLabel": [Function],
        "text": "Option A",
      },
      Object {
        "key": "B",
        "onRenderLabel": undefined,
        "text": "Option B",
      },
    ]
  }
  selectedKey="B"
/>
`;

exports[`<ChoiceGroup /> with minimal props renders as expected 1`] = `
<ChoiceGroup
  className="y-choice-group"
  onChange={[Function]}
  options={
    Array [
      Object {
        "key": "A",
        "onRenderLabel": [Function],
        "text": "Option A",
      },
      Object {
        "key": "B",
        "onRenderLabel": undefined,
        "text": "Option B",
      },
    ]
  }
/>
`;

exports[`<ChoiceGroup /> with minimal props sublabel renders as expected 1`] = `
<span
  className="ms-Label"
>
  <Text
    bold={true}
  >
    Option A:
  </Text>
  <Text>
    Â 
    Sublabel for A
  </Text>
</span>
`;

exports[`<ChoiceGroup /> with minimal props when a change is triggered with an option renders as expected 1`] = `
<ChoiceGroup
  className="y-choice-group"
  onChange={[Function]}
  options={
    Array [
      Object {
        "key": "A",
        "onRenderLabel": [Function],
        "text": "Option A",
      },
      Object {
        "key": "B",
        "onRenderLabel": undefined,
        "text": "Option B",
      },
    ]
  }
/>
`;

exports[`<ChoiceGroup /> with minimal props when a change is triggered with an option that is \`undefined\` renders as expected 1`] = `
<ChoiceGroup
  className="y-choice-group"
  onChange={[Function]}
  options={
    Array [
      Object {
        "key": "A",
        "onRenderLabel": [Function],
        "text": "Option A",
      },
      Object {
        "key": "B",
        "onRenderLabel": undefined,
        "text": "Option B",
      },
    ]
  }
/>
`;
